datafusion/common/src/tree_node.rs

```diff
@@ -377,32 +583,408 @@ pub trait ConcreteTreeNode: Sized {
+            // transformed children.
+            Ok(Transformed::new(t2.data, t.transformed, t.tnr))
         } else {
-            Ok(new_self)
+            Ok(Transformed::no(new_self))
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::tree_node::{
+        Transformed, TransformedIterator, TreeNode, TreeNodeRecursion, TreeNodeRewriter,
+        TreeNodeVisitor,
+    };
+    use crate::Result;
+
+    struct TestTreeNode<T> {
+        children: Vec<TestTreeNode<T>>,
+        data: T,
+    }
+
+    impl<T> TestTreeNode<T> {
+        fn new(children: Vec<TestTreeNode<T>>, data: T) -> Self {
+            Self { children, data }
+        }
+    }
+
+    impl<T> TreeNode for TestTreeNode<T> {
+        fn apply_children<F>(&self, f: &mut F) -> Result<TreeNodeRecursion>
+        where
+            F: FnMut(&Self) -> Result<TreeNodeRecursion>,
+        {
+            let mut tnr = TreeNodeRecursion::Continue;
+            for child in &self.children {
+                tnr = f(child)?;
+                handle_visit_recursion!(tnr);
+            }
+            Ok(tnr)
+        }
+
+        fn map_children<F>(self, f: F) -> Result<Transformed<Self>>
+        where
+            F: FnMut(Self) -> Result<Transformed<Self>>,
+        {
+            Ok(self
+                .children
+                .into_iter()
+                .map_till_continue_and_collect(f)?
+                .map_data(|new_children| Self {
+                    children: new_children,
+                    ..self
+                }))
+        }
+    }
+
+    fn new_test_tree<'a>() -> TestTreeNode<&'a str> {
+        let node_a = TestTreeNode::new(vec![], "a");
+        let node_b = TestTreeNode::new(vec![], "b");
+        let node_d = TestTreeNode::new(vec![node_a], "d");
+        let node_c = TestTreeNode::new(vec![node_b, node_d], "c");
+        let node_e = TestTreeNode::new(vec![node_c], "e");
+        let node_h = TestTreeNode::new(vec![], "h");
+        let node_g = TestTreeNode::new(vec![node_h], "g");
+        let node_f = TestTreeNode::new(vec![node_e, node_g], "f");
+        let node_i = TestTreeNode::new(vec![node_f], "i");
+        TestTreeNode::new(vec![node_i], "j")
+    }
+
+    fn all_visits<'a>() -> Vec<&'a str> {
+        vec![
+            "f_down(j)",
+            "f_down(i)",
+            "f_down(f)",
+            "f_down(e)",
+            "f_down(c)",
+            "f_down(b)",
+            "f_up(b)",
+            "f_down(d)",
+            "f_down(a)",
+            "f_up(a)",
+            "f_up(d)",
+            "f_up(c)",
+            "f_up(e)",
+            "f_down(g)",
+            "f_down(h)",
+            "f_up(h)",
+            "f_up(g)",
+            "f_up(f)",
+            "f_up(i)",
+            "f_up(j)",
+        ]
+    }
+
+    fn f_down_jump_visits<'a>() -> Vec<&'a str> {
+        vec![
+            "f_down(j)",
+            "f_down(i)",
+            "f_down(f)",
+            "f_down(e)",
+            "f_down(g)",
```

Comment:
Did you intentionally pass to "f_down(g)" after jump from "f_down(e)" ? I believe it will be more useful and expected to apply "f_up(e)" before jumping to "f_down(g)". What do you think about that, can you modify it in that way without much effort?

Reply:
Actually I did. Because does it make any sense to call `f_up(e)` when `f_down(e)` returned `Jump`? I don't think it does, as you can do everything in `f_down(e)` (transform the node for example) besides returning `Jump`.
This behavior is aligned with the current `VisitRecursion::Skip` so if we changed this behaviour and call `f_up(e)` too, then some of the `TreeNodeVisitor` logics might need to change...